### ISSUE TITLE
build tool plugin fail on Windows just after launch

### ISSUE THREAD

**Issue by ericsink** [apple/swift-package-manager#5668]
PR #5664 seems to fix the Path problem, but then I bump into the *other* problem mentioned in my original forum post:

https://forums.swift.org/t/known-case-of-a-working-build-tool-plugin-on-windows/58825

I had been hoping this other problem was somehow also related to the Path, but it appears to be distinct.  In a nutshell, it seems that the build tool plugin always dies returning 1033:

```
error: plugin process ended by an uncaught signal: 1033 <command: "C:\Users\eric\dev\swiftpm-buildtool-plugin-examples\Example 1 - SwiftGen\.build\plugins\cache\SwiftGenPlugin.exe">, <output:
>
```

Note that the build tool plugin code, as long it compiles, does not affect the error behavior:

```
@main
struct SwiftGenPlugin: BuildToolPlugin {

    func createBuildCommands(context: PluginContext, target: Target) throws -> [Command] {
        // TODO nothing I put here seems to matter
        // for example, printing something to stderr here does not get included in the error message
        return []
    }
```

Bottom line: AFAICT, unless I am doing something wrong, build tool plugins still don't work on Windows at all.

_Originally posted by @ericsink in https://github.com/apple/swift-package-manager/issues/5664#issuecomment-1185784583_

**Comment by ericsink** [comment]
Comment posted by @compnerd on #5664:

The failure that you are seeing is interesting. I suspect that it is something different - likely a bad expectation somewhere or argument passing is triggering this. Quickly checking the binary that was generated, running it locally does indicate that it immediately fails as well ... though its mode of failure is different but interesting. The error level is -1073740791. Using 2s complement, we see that this is 0xc0000409. That indicates that it is a STACK_BUFFER_UNDERRUN. A quick look at the stack trace shows the following:

```
 # Child-SP          RetAddr               Call Site
00 000000dc`68eff520 00007ffd`bc7c24b1     ucrtbase!invoke_watson+0x18
01 000000dc`68eff550 00007ffd`bc7c2379     ucrtbase!_invalid_parameter+0x12d
02 000000dc`68eff590 00007ffd`bc82476f     ucrtbase!invalid_parameter_noinfo+0x9
03 000000dc`68eff5d0 00007ffd`5e712474     ucrtbase!setvbuf+0x8f
04 000000dc`68eff630 00007ffd`5f99b0e6     PackagePlugin!$s13PackagePlugin0B0PAAE4mainyyYaKFZ+0x344
05 000000dc`68eff960 00007ffd`5f99b8a7     swift_Concurrency!$ss27AsyncThrowingFilterSequenceV8IteratorV10isIncludedySb7ElementQzYaKcvg+0x2c16
06 000000dc`68eff9a0 00007ffd`5f92ab50     swift_Concurrency!swift_job_run+0x57
07 000000dc`68effa00 00007ffd`5f92a92c     dispatch!dispatch_assert_queue_barrier+0x510
08 000000dc`68effa60 00007ffd`5f9354e4     dispatch!dispatch_assert_queue_barrier+0x2ec
09 000000dc`68effad0 00007ffd`bc7c1bb2     dispatch!dispatch_prohibit_transition_to_multithreaded+0x2534
0a 000000dc`68effb90 00007ffd`be877034     ucrtbase!thread_start<unsigned int (__cdecl*)(void *),1>+0x42
0b 000000dc`68effbc0 00007ffd`befe2651     KERNEL32!BaseThreadInitThunk+0x14
0c 000000dc`68effbf0 00000000`00000000     ntdll!RtlUserThreadStart+0x21
```

A closer look at frame 3:

```
03 000000dc`68eff5d0 00007ffd`5e712474     : 00007ffd`bc88f4e8 00000000`00000000 00000000`00000040 00000000`00000000 : ucrtbase!setvbuf+0x8f
```

indicates that somehow we ended up with `setvbuf(0x7ffdbc88f4e8, NULL, 0x40, 0);` which is triggering Dr Watson ...

**Comment by ericsink** [comment]
Indeed, I see only one `setvbuf` call in the code:

https://github.com/apple/swift-package-manager/blob/e25a590dc455baa430f2ec97eacc30257c172be2/Sources/PackagePlugin/Plugin.swift#L87

The whole sequence there at the beginning of plugin `main()` is manipulating file descriptors for stdout and stderr to redirect things.  Something there isn't working properly on Windows.  Maybe it's just the `setvbuf` call, but I'm slightly suspicious of the code above it as well.

**Comment by ericsink** [comment]
Makes sense.  I've been trying to setup a build environment so I could perhaps be a bit more helpful.  Not quite there yet...

**Comment by compnerd** [comment]
Verified with some local hacks, the `setvbuf` call is the problem.  However, disabling that yields another issue - `swiftgen` is unavailable as it expects to have a binary artifact for that which is Darwin only.  That sounds like this is just an issue of improving the handling for the standard console handles.

**Comment by ericsink** [comment]
The swiftgen unavailable problem would be expected, right? Sounds like the same error I see when trying to run that example unmodified on Linux.

I don't expect swiftgen to actually work without providing a binary for the platform.

**Comment by ericsink** [comment]
But if I can somehow get a build setup going (no luck so far), or download a build with your setvbuf fix, I can verify a (trivial) real plugin.

**Comment by compnerd** [comment]
I'll upload a fix and kick off another build.

**Comment by compnerd** [comment]
@ericsink - https://ci-external.swift.org/job/swift-PR-build-toolchain-windows/137/artifact/build/artifacts/installer.exe ... I believe should address the crash that you are seeing.

**Comment by ericsink** [comment]
@compnerd -- confirming:  So far, that build works as I would expect.  The plugin runs.  If I configured a binary in `info.json` for `x86_64-unknown-windows-msvc` then it gets invoked.

I plan to do a bit more thorough test with a still-minimal plugin that actually does something, but the fixes look great so far.  Thanks!
